File: agent/tools.py
from typing import Optional, Dict, Any
from utilities.neo4j_utils import Neo4jManager
from agent.config import BROWSER_DOMAIN, BROWSER_USERNAME, BROWSER_PASSWORD

# グローバルNeo4jマネージャーインスタンス（後で初期化）
neo4j_manager: Optional[Neo4jManager] = None

import asyncio
from playwright.async_api import async_playwright

async def execute_workflow_async(workflow):
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=False)
        context = await browser.new_context()
        page = await context.new_page()
        
        # Login if needed (assuming from config)
        await page.goto(BROWSER_DOMAIN)
        await page.wait_for_load_state('networkidle', timeout=30000)
        login_input = await page.query_selector('input[name="emailOrUsername"]')
        if login_input:
            await login_input.fill(BROWSER_USERNAME)
        password_input = await page.query_selector('input[type="password"]')
        if password_input:
            await password_input.fill(BROWSER_PASSWORD)
        submit_button = await page.query_selector('button.login')
        if submit_button:
            await submit_button.click()
        await page.wait_for_load_state('networkidle', timeout=30000)
        await page.wait_for_timeout(5000)  # Additional wait for dynamic content
        
        results = []
        snapshots = []  # To collect ARIA snapshots
        
        try:
            for i, step in enumerate(workflow):
                action = step.get('action')
                try:
                    if action == 'goto':
                        url = step.get('url')
                        if url:
                            await page.goto(url)
                            results.append(f"Navigated to {url}")
                            await page.wait_for_load_state('networkidle', timeout=30000)
                        else:
                            results.append("Invalid goto step: missing url")
                    elif action in ['click', 'input', 'press']:
                        role = step.get('role')
                        name = step.get('name')
                        if role and name:
                            locator = page.get_by_role(role, name=name, exact=True)
                            await locator.wait_for(state="visible", timeout=30000)
                            
                            if action == 'click':
                                await locator.click()
                                results.append(f"Clicked {role}: {name}")
                            elif action == 'input':
                                text = step.get('text')
                                if text:
                                    await locator.fill(text)
                                    results.append(f"Input {text} into {role}: {name}")
                            elif action == 'press':
                                key = step.get('key')
                                if key:
                                    await locator.press(key)
                                    results.append(f"Pressed {key} on {role}: {name}")
                            # Add more actions as needed
                            await page.wait_for_load_state('networkidle', timeout=30000)
                        else:
                            results.append("Invalid step: missing role or name")
                    else:
                        results.append(f"Unknown action: {action}")
                    
                    # Capture ARIA snapshot after each step
                    snapshot = await page.accessibility.snapshot()
                    snapshots.append(f"ARIA Snapshot after step {i+1}: {snapshot}")
                
                except Exception as step_error:
                    # Capture snapshot on error
                    error_snapshot = await page.accessibility.snapshot()
                    results.append(f"Error in step {i+1}: {str(step_error)}\nError ARIA Snapshot: {error_snapshot}")
                    snapshots.append(f"Error ARIA Snapshot for step {i+1}: {error_snapshot}")
                    raise  # Re-raise to handle outer try-except
            
            # Final snapshot after all steps
            final_snapshot = await page.accessibility.snapshot()
            snapshots.append(f"Final ARIA Snapshot: {final_snapshot}")
        
        except Exception as e:
            results.append(f"Workflow execution failed: {str(e)}")
        
        finally:
            await browser.close()
        
        return "\n".join(results) + "\n\nSnapshots:\n" + "\n".join(snapshots) + "\nWorkflow executed."

def execute_workflow(workflow: list[dict]) -> str:
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
    return loop.run_until_complete(execute_workflow_async(workflow))

def run_cypher(query: str) -> str:
    """
    Neo4jでCypherクエリを実行
    
    Args:
        query: 実行するCypherクエリ
    
    Returns:
        クエリ実行結果の文字列表現
    """
    global neo4j_manager
    
    if not neo4j_manager:
        return "エラー: Neo4jに接続されていません"
    
    try:
        print(f"\n🔍 実行するクエリ:\n{query}\n")
        results = neo4j_manager.execute_cypher(query)
        
        if not results:
            return "結果: データが見つかりませんでした"
        
        # 結果を見やすく整形
        output = []
        for i, record in enumerate(results[:20]):  # 最大20件まで表示
            record_dict = dict(record)
            output.append(f"レコード {i+1}: {record_dict}")
        
        if len(results) > 20:
            output.append(f"\n... 他 {len(results) - 20} 件のレコードがあります")
        
        return "\n".join(output)
        
    except Exception as e:
        return f"クエリ実行エラー: {str(e)}" 

File: src/agent/tools.ts
import { chromium } from 'playwright';
import type { Driver } from 'neo4j-driver';
import { createDriver, closeDriver } from '../utilities/neo4j.js';
import { getSnapshotForAI } from '../utilities/snapshots.js';
import { findRoleAndNameByRef } from '../utilities/text.js';

export async function runCypher(query: string): Promise<string> {
  const uri = process.env.AGENT_NEO4J_URI;
  const user = process.env.AGENT_NEO4J_USER;
  const password = process.env.AGENT_NEO4J_PASSWORD;
  if (!uri || !user || !password) return 'エラー: Neo4j接続情報(AGENT_NEO4J_URI/AGENT_NEO4J_USER/AGENT_NEO4J_PASSWORD)が未設定です';

  let driver: Driver | null = null;
  try {
    driver = await createDriver(uri, user, password);
    const session = driver.session();
    try {
      const res = await session.run(query);
      const records = res.records.map((r) => r.toObject());
      if (!records.length) return '結果: データが見つかりませんでした';
      const lines: string[] = [];
      records.slice(0, 20).forEach((rec, i) => lines.push(`レコード ${i + 1}: ${JSON.stringify(rec)}`));
      if (records.length > 20) lines.push(`\n... 他 ${records.length - 20} 件のレコードがあります`);
      return lines.join('\n');
    } finally {
      await session.close();
    }
  } catch (e: any) {
    return `クエリ実行エラー: ${String(e?.message ?? e)}`;
  } finally {
    await closeDriver(driver);
  }
}

type WorkflowStep =
  | { action: 'goto'; url: string }
  | { action: 'click'; ref: string }
  | { action: 'input'; ref: string; text: string }
  | { action: 'press'; ref: string; key: string };

export async function executeWorkflow(workflow: WorkflowStep[]): Promise<string> {
  const headful = String(process.env.AGENT_HEADFUL ?? 'false').toLowerCase() === 'true';
  const browser = await chromium.launch({ headless: !headful });
  const context = await browser.newContext();
  const page = await context.newPage();
  const results: string[] = [];
  const snapshots: string[] = [];
  try {
    // Optional pre-login flow to align with Python implementation
    const domain = process.env.AGENT_BROWSER_DOMAIN;
    const username = process.env.AGENT_BROWSER_USERNAME;
    const password = process.env.AGENT_BROWSER_PASSWORD;
    if (domain && username && password) {
      try {
        console.log(`[Login] ${domain} にアクセスしてログインを試行します`);
        await page.goto(domain);
        await page.waitForLoadState('networkidle').catch(() => {});
        const loginInput = await page.$('input[name="emailOrUsername"]');
        if (loginInput) await loginInput.fill(username);
        const pwInput = await page.$('input[type="password"]');
        if (pwInput) await pwInput.fill(password);
        const submit = await page.$('button.login');
        if (submit) await submit.click();
        await page.waitForLoadState('networkidle').catch(() => {});
        await page.waitForTimeout(5000).catch(() => {});
        results.push('Pre-login succeeded');
      } catch (e: any) {
        results.push(`Pre-login failed: ${String(e?.message ?? e)}`);
      }
    }

    // 提供: 現ページのARIAスナップショットに [ref=eXX] を付与する関数を注入
    async function installRefSnapshotProvider() {
      const makeSnapshot = async (): Promise<string> => {
        const tree: any = await page.accessibility.snapshot().catch(() => ({}));
        const lines: string[] = [];
        let counter = 1;
        const walk = (node: any, depth: number) => {
          if (!node || typeof node !== 'object') return;
          const role = String(node.role ?? 'generic');
          const name = typeof node.name === 'string' ? node.name : undefined;
          const ref = `e${counter++}`;
          const label = name && name.trim().length > 0 ? `${role} "${name}" [ref=${ref}]` : `${role} [ref=${ref}]`;
          lines.push(`${'  '.repeat(depth)}- ${label}`);
          const children: any[] = Array.isArray(node.children) ? node.children : [];
          for (const c of children) walk(c, depth + 1);
        };
        walk(tree, 0);
        return lines.join('\n');
      };
      (page as any)._snapshotForAI = makeSnapshot;
    }

    async function resolveLocatorByRef(ref: string) {
      // 必ずプロバイダをセットしてから参照解決
      await installRefSnapshotProvider();
      const snapText = await getSnapshotForAI(page);
      const roleName = findRoleAndNameByRef(snapText, ref);
      if (!roleName) throw new Error(`ref=${ref} に対応する要素が見つかりません (現在のARIAスナップショット)`);
      const { role, name } = roleName;
      const locator = name && name.trim().length > 0
        ? page.getByRole(role as any, { name, exact: true } as any)
        : page.getByRole(role as any);
      // 要素の存在確認
      await locator.first().waitFor({ state: 'visible', timeout: 30000 });
      return { locator, role, name } as const;
    }

    for (let i = 0; i < workflow.length; i += 1) {
      const step = workflow[i]!;
      try {
        if (step.action === 'goto') {
          await page.goto(step.url);
          await page.waitForLoadState('networkidle').catch(() => {});
          results.push(`Navigated to ${step.url}`);
        } else if (step.action === 'click') {
          const { locator, role, name } = await resolveLocatorByRef(step.ref);
          await locator.first().click();
          await page.waitForLoadState('networkidle').catch(() => {});
          results.push(`Clicked ref=${step.ref} (${role}${name ? `: ${name}` : ''})`);
        } else if (step.action === 'input') {
          const { locator, role, name } = await resolveLocatorByRef(step.ref);
          await locator.first().fill(step.text);
          results.push(`Input into ref=${step.ref} (${role}${name ? `: ${name}` : ''}) -> ${step.text}`);
        } else if (step.action === 'press') {
          const { locator, role, name } = await resolveLocatorByRef(step.ref);
          await locator.first().press(step.key);
          results.push(`Pressed ${step.key} on ref=${step.ref} (${role}${name ? `: ${name}` : ''})`);
        } else {
          results.push(`Unknown action: ${(step as any).action}`);
        }
      } catch (e: any) {
        const snap = await page.accessibility.snapshot().catch(() => ({}));
        const textSnap = await getSnapshotForAI(page).catch(() => '');
        results.push(`Error in step ${i + 1}: ${String(e?.message ?? e)}\nError ARIA Snapshot: ${JSON.stringify(snap)}`);
        snapshots.push(`Error ARIA Snapshot for step ${i + 1}: ${JSON.stringify(snap)}`);
        if (textSnap) snapshots.push(`Error Text Snapshot with refs for step ${i + 1}:\n${textSnap}`);
        break;
      }
      const snap = await page.accessibility.snapshot().catch(() => ({}));
      const textSnap = await getSnapshotForAI(page).catch(() => '');
      snapshots.push(`ARIA Snapshot after step ${i + 1}: ${JSON.stringify(snap)}`);
      if (textSnap) snapshots.push(`Text Snapshot with refs after step ${i + 1}:\n${textSnap}`);
    }
    const finalSnap = await page.accessibility.snapshot().catch(() => ({}));
    const finalTextSnap = await getSnapshotForAI(page).catch(() => '');
    snapshots.push(`Final ARIA Snapshot: ${JSON.stringify(finalSnap)}`);
    if (finalTextSnap) snapshots.push(`Final Text Snapshot with refs:\n${finalTextSnap}`);
  } finally {
    await browser.close();
  }
  return results.join('\n') + '\n\nSnapshots:\n' + snapshots.join('\n') + '\nWorkflow executed.';
}

export type ToolUseInput =
  | { name: 'run_cypher'; input: { query: string }; toolUseId: string }
  | { name: 'execute_workflow'; input: { workflow: WorkflowStep[] }; toolUseId: string };


